javaharness/java/arcs/android/demo/service/NotificationRenderer.java

```diff
@@ -53,25 +57,44 @@ public boolean render(PortableJson packet) {
         new Notification.Builder(context, CHANNEL_ID)
             .setSmallIcon(R.drawable.baseline_notification_important_black_18)
             .setContentTitle(title)
-            .setContentText(text)
-            .setAutoCancel(true);
+            .setContentText(text);
 
     String outputSlotId = packet.getString(OUTPUT_SLOT_ID_FIELD);
     // TODO(mmandlis): refactor to a generic method usable by other renderers as well.
-    if (content.hasKey(HANDLER_FIELD)) {
-      String handler = content.getString(HANDLER_FIELD);
-
-      Intent notificationIntent = new Intent(context, ArcsService.class);
-      notificationIntent.putExtra(ArcsService.INTENT_REFERENCE_ID_FIELD, outputSlotId);
-      notificationIntent.putExtra(
-          ArcsService.INTENT_EVENT_DATA_FIELD,
-          jsonParser.stringify(jsonParser.emptyObject().put(HANDLER_FIELD, handler)));
+    if (content.hasKey(TAP_HANDLER_FIELD)) {
+      // Construct intent for notification tap.
+      String handler = content.getString(TAP_HANDLER_FIELD);
+
+      Intent intent = new Intent(context, ArcsService.class);
+      intent.setAction(outputSlotId);
+      intent.putExtra(ArcsService.INTENT_REFERENCE_ID_FIELD, outputSlotId);
+      intent.putExtra(
+        ArcsService.INTENT_EVENT_DATA_FIELD,
+        jsonParser.stringify(
+          jsonParser.emptyObject().put(HANDLER_FIELD, handler)));
       PendingIntent pendingIntent =
-          PendingIntent.getService(
-              context, 0, notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+        PendingIntent.getService(
+          context, REQUEST_CODE_TAP, intent, PendingIntent.FLAG_UPDATE_CURRENT);
       builder.setContentIntent(pendingIntent);
     }
 
+    if (content.hasKey(DISMISS_HANDLER_FIELD)) {
+      // Construct intent for notification dismiss.
+      String handler = content.getString(DISMISS_HANDLER_FIELD);
+
+      Intent intent = new Intent(context, ArcsService.class);
```

Comment:
it seems worth refactoring the intent creation into a helper method.

Reply:
Done.